Sam.jaberi

samnjab opened the pull request:
For the purposes of making this PR quickly, everything accomplished in this PR are explained in the commits and I'm not going to repeat them.

samnjab committed:
Created a LandingInputBoundary with actions to be performed on the landing page.

samnjab committed:
Created RegionInputData to encapsulate information about the selected region and getter functions.

samnjab committed:
Created LandingInteractor that encapsulates a RegionDataAccessInterface of all the available regions and a LandingOutputBoudary object as a presenter. It 1. implements the selectRegion method, handles appropriate exceptions related to selecting a region 2. handles view transitions to signup and login views.

samnjab committed:
created a Region interface in entities.

samnjab committed:
modified the structure of the region interface to match the region blueprint.

samnjab committed:
Trimmed down the structure of the abstract class AbstractRegion, and refactored all the region entities into a separate package.

samnjab committed:
Got rid of the region interface and abstractRegion for a cleaner architecture.

samnjab committed:
Implemented RegionOutputData and added prepareFailView conditions to LandingInteractor.

samnjab committed:
Implemented LandingInputBoundary interface that defines the result of the actions performed on the landing page.

samnjab committed:
Implemented RegionDataAccessInterface that defines methods for searching through and retrieving from the region database.

samnjab committed:
Added Identifier, Summary, and Address entities.

samnjab committed:
Added Area entity. Modified Region entity to encapsulate Area properties. Deleted Country, State, City, County classes as they are all Area's now.

samnjab committed:
Cleaned up usecases, replaced RegionInputData with AddressInputData and made corresponding changes to LandingController and LandinInteractor.

samnjab committed:
Created AreaOutputData which is to replace RegionOutputData.

samnjab committed:
Changes to LandingPresenter in progress.

samnjab committed:
Summary accepts String values for condition and propType and maps them into int values now.
Property constructor handles null pointers.

samnjab committed:
Identifier validates attomId, geoId, and geoIdV4. geoIdV4 map must contain all the following keys "ST", "CO", "CS", "N1", and "ZI".

samnjab committed:
Got rid of equal and toHash method on Summary and Address.

samnjab committed:
AddressInputData now includes Area zipCode as a field.

samnjab committed:
Created AreaDataAccessInteface that provides methods to search in Area database.

samnjab committed:
LandingPresenter implements prepare success views for Area and list of areas. LandingOutputBoundary contains all the methods required to present results.

samnjab committed:
- AreaListOutputData DTO implemented.
- AreaDataInterface edited comments and method names.
- LandingInputBoundary includes improved methods fetchAreas and selectArea.
_LandingInteractor correctly handles fetchAreas and selectArea usecases.

samnjab committed:
- AttomClient is the API contact points and handles all the calls. All methods return JasonNode

samnjab committed:
- Added findAreasByNameAndType to AreaDataAccessInterface.

samnjab committed:
- Added autoCompleteByName method that fetches and returns a list of Areas with .type = type and matching .name and partialName.

samnjab committed:
- fixed a ; issue.
- implemented autoCompleteByName logic.

samnjab committed:
- first draft LandingViewModel
- first draft LandingState

samnjab committed:
- AreaStore acts as area cache, where the fetched subareas will reside.

samnjab committed:
- Updated AddressInputData.

samnjab committed:
- AreaDataAccessObject implements AreaDataAccessInterface with AreaStore object as a field.
- Updated getSubAreas() method in AreaDataAccessInterface to accept param type.

samnjab committed:
- delegated findAreasByNameAndType() method to the interactor, removed it from AreaStore.
- AttomClient now includes a method fetchSubAreas to search for subareas of an area.

samnjab committed:
- methods in AttomClient return specific parts of the response.body() as a JsonNode now.
- Added a custom exception class to the infrastructure layer to handle api request fails.

samnjab committed:
- Implemented getCountries() and getSubAreas() in AreaDataAccessObject.
- Got rid of getArea() and existsByGeoId() methods in the AreaDataAccessInterface.

samnjab committed:
- Implemented fetchStates() in AttomClient.

samnjab committed:
- Added an overload conditionals to Area.name and Area.nameCode initializations to handle areas with no name or name code.
- Added an overload constructor to Area.
- Added getCountries method to AreaDataAccessInterface.
- Added postalCode as a field to Address.
- added postalCode to constructAddress method in AddressInputData.

samnjab committed:
- added selectArea, cacheAreas to AreaDataAccessObject
-  Added AreaStore() constructor that initializes areaCache and areaMap.
- Added storeArea() method to AreaStore.

samnjab committed:
- removed geoId field and getter from Identifier
- updated LandingInputBoundary.
- Draft of LandingInteractor (incomplete).

samnjab committed:
- Overloaded Property constructor for API exception handling.

samnjab committed:
- Added PropertyDaraAccessInterface.
- Added PropertyDataAccessObject.
- Overloaded Property constructor for API exception handling.

samnjab committed:
Merge remote-tracking branch 'origin/sam.jaberi' into sam.jaberi

samnjab committed:
- Overloaded Property constructor for API exception handling.

samnjab committed:
- fixed my pom file.
compiler 22 -> 23
removed jackson 2.17.

samnjab committed:
- fetchSalesComparables method added to AttomClient that returns the whole response.body() for now. Dataa needs parsing.
- made ComparableSaleDto to hold sales info for the comaprator usecase.
- Added propertyCache field to PropertyDataAccessObject.
- overloaded Address, Identifier, and Summary constructors to make an empty objects.
- AddressNotFound Exception
- storeArea() method first clears the previously held data at areaType.
- LandingInputBoundary contains no method with return.

ranb0m commented on src/main/java/quotify_app/adapters/landing/LandingController.java:
We'd need to add the functionality on the landing page of being able to navigate to the user profile for logged in users.
This spans the landing usecase, I'll just leave one reference here.

You can take inspiration from the other components and how they add this functionality, it should be exactly the same here too

ranb0m commented on src/main/java/quotify_app/entities/regionEntities/Area.java:
nothing wrong with the file, just curious: asides from the landing/inputting information, do we elsewhere make use of the area/AreaStore stuff?

samnjab replied:
Not directly. The AreaDataAccessObject calls the prediction model "API" instead of Attom in a number of its methods, like getCountries and getSubAreas, to get available areas and cache them into its AreaStore instance. 
I can't recall any other parts of the program utilizing Area in particular in its current edition.

ranb0m commented on src/main/java/quotify_app/usecases/landing/LandingInteractor.java:
```diff
@@ -0,0 +1,93 @@
+            final AreaListOutputData outputData = new AreaListOutputData(areaList, type, false);
+            landingPresenter.prepareAreaListSuccessView(outputData);
+        }
+        catch (Exception e) {
```
does checkstyle not complain about this? I have a feeling it might later on -- can try adding a more specific type of exception here

samnjab replied:
Absolutely, I'll fix all instances of catching Exception e.

ranb0m commented on src/main/java/quotify_app/data_access/AttomClient.java:
you mentioned this file needed to be cleaned up, I'll just leave a reminder note

ranb0m reviewed (changes requested):
Left a few comments: the main thing I already mentioned, to try and add to the appbuilder and to make a factory the builds the relevant landing stuff.

Another thing, not urgent since I'm not done the user profile yet, but a way to navigate to the user profile from the landing for logged in users, also left a comment for that.

Asides from that, looking good! great code style

samnjab committed:
- refactored AttomClient into helper methods to avoid code repetition.
- overloaded constructor for ApiRequestException.
- Added ClientRequestException to group IO and Interrupted Exceptions.

samnjab committed:
- LandingInteractor handles custom exceptions.
- Added fetchCountries to interactor.
- prepareErrorView handles general program errors in the presenter.
- Added getCurrentProperty and setCurrentProperty methods to PropertyDataAccessInterface.

ranb0m commented on pom.xml:
```diff
@@ -9,8 +9,10 @@
 
     <properties>
-        <maven.compiler.source>22</maven.compiler.source>
-        <maven.compiler.target>22</maven.compiler.target>
```
why are we switching to java 23 here? do we know this doesn't introduce any weird behaviour with the other dependencies/everyone else's setup?

ranb0m reviewed (approved):
Requested changes taken care of! ready to resolve pom conflict from main and merge.

ranb0m committed:
Merge branch 'main' into sam.jaberi
